### DISCUSSION TITLE
Use CDI interceptor for @Nested jUnit tests

### DISCUSSION THREAD

**Discussion by HerrDerb** [quarkusio/quarkus#36635] [Q&A] [answered]
I have a `@QuarkusTest` class which contains quite a few test. For a better overview, I am organizing them with inner `@Nested` classes.
Now I also use an annotation bound CDI Interceptor on the main test class. The interceptor for all methods which directly implemented directly in the main class. All test methods inside a inner `@Nested` class are ignored.
I guess this is, as a `@Nested` class is not part of the CDI. This at least is also the message I get, when I try to put my annotation directly on a inner class.

Is it possible to use a CDI interceptor on a test class that also works for its inner `@Nested` test classes?

[a comment by an automated account is not shown]

**Comment by Ladicek** [comment]
No. The interceptor only works because the test class is treated as a bean and the test instance is obtained from the CDI container. This clearly is not the case for `@Nested` inner classes, and it cannot be -- the `@Nested` classes are supposed to be non-`static`, and non-`static` member classes cannot be beans.

**Reply by HerrDerb** [reply]
@Ladicek Thanks for the quick answer. So there is currently no "Quarkus" way to solve this?

**Reply by manovotn** [reply]
Ladislav is correct.
I'd just add that this is also backed by [CDI specification](https://jakarta.ee/specifications/cdi/4.0/jakarta-cdi-spec-4.0#what_classes_are_beans) if you want to read deeper into it.

**Reply by Ladicek** [reply]
You might be able to use the test lifecycle callbacks: https://quarkus.io/guides/getting-started-testing#enrichment-via-quarkustestcallback I never tried it, but these callbacks are based on the JUnit Jupiter extension API, so they should work.

**Reply by manovotn** [reply, marked as answer]
> So there is currently no "Quarkus" way to solve this?

I don't think there is, at least in terms of CDI interception. Nested classes cannot become beans; the biggest issue is the way they are instantiated and the affect it would have in terms of bean (re)creation and proxyability.
If you really need CDI interception of test instances, avoid using `@Nested` and organize tests differently.

**Reply by HerrDerb** [reply]
@Ladicek I already had a look at them. For my use case, I need to intercept and "wrap" the invocation (similar to TestTransaction)

Thank your for the clarification
